build_mt.py

```diff
@@ -1,599 +1,827 @@
+#!/usr/bin/env python3
+'''
+This is the first script in the "lightcone halo" pipeline.  The goal of this script is to use merger
+tree information to flag halos that intersect the lightcone and make a unique determination of which
+halo catalog epoch from which to draw the halo.
+
+Usage
+-----
+$ ./build_mt.py --help
+'''
+
+import sys
 import glob
+import time
+import gc
+import os
+from pathlib import Path
+
 import asdf
 import numpy as np
-import sys
 from scipy.interpolate import interp1d
 import matplotlib.pyplot as plt
-import numpy.lib.recfunctions as rfn
-import time
-import gc
-import os
+import numba as nb
 
 from tools.merger import simple_load, get_slab_halo, extract_superslab
 
+# these are probably just for testing; should be removed for production
+DEFAULTS = {}
+DEFAULTS['sim_name'] = "AbacusSummit_highbase_c000_ph100"  # AbacusSummit_base_c000_ph006
+DEFAULTS['z_start'] = 0.3  # 0.8#0.5
+DEFAULTS['z_stop'] = 0.65  # 1.25#0.8#0.5
+
+
 # reorder in terms of their slab number
 def reorder_by_slab(fns):
-    i_sort = np.argsort(extract_superslab(fns))
-    tmp = fns.copy()
-    for i in range(len(fns)):
-        tmp[i] = fns[i_sort[i]]
-    return tmp
+    return sorted(fns, key=extract_superslab)
+
 
 # read redshifts from merger tree files
 def get_zs_from_headers(snap_names):
     zs = np.zeros(len(snap_names))
     for i in range(len(snap_names)):
         snap_name = snap_names[i]
-        f = asdf.open(snap_name,lazy_load=True, copy_arrays=True)
-        z = np.float(f.tree['header']['Redshift'])
-        f.close()
-        zs[i] = z
+        with asdf.open(snap_name) as f:
+            zs[i] = np.float(f["header"]["Redshift"])
     return zs
 
-def distance_period(x0, x1, dimensions):
-    delta = np.abs(x0 - x1)
-    delta = np.where(delta > 0.5 * dimensions, delta - dimensions, delta)
-    return np.sqrt((delta ** 2).sum(axis=-1))
 
-def distance(x0, x1):
-    return np.sqrt(np.sum((x0-x1)**2,axis=1))
+@nb.njit
+def dist(pos1, pos2, L=None):
+    '''
+    Calculate L2 norm distance, optionally including periodicity
+    '''
+    assert pos1.shape == pos2.shape
+    N, nd = pos1.shape
+    dist = np.empty(N, dtype=pos1.dtype)
+    
+    for i in range(N):
+        delta = 0.
+        for j in range(nd):
+            dx = pos1[i][j] - pos2[i][j]
+            if L is not None:
+                if dx >= L/2:
+                    dx -= L
+                elif dx < -L/2:
+                    dx += L
+            delta += dx*dx
+        dist[i] = np.sqrt(delta)
+    return dist
+
```

Comment:
Numpy isn't particularly good at doing these distance calculations; it makes a bunch of temporary arrays and loops over the inputs many times.  Numba is good at this kind of element-oriented work.  The numba version runs about 5x faster.  This will get used many times, so it's worth the longer implementation!

Also, note this includes both the periodic and non-periodic behavior, with the `L` arg.

Reply:
wow! am kinda surprised that element-wise numba is faster than this "vector" numpy operation, but that's super useful to know!

Reply:
One slightly confusing thing here is that I am computing this with respect to just one point, the observer location

Reply:
Oh, right, then the assertion at the top of this function that the inputs must be the same shape should change.  Then the `pos2` indexing will need to know about that... let me fix it tomorrow.

On the speed, it's important to remember that Numba is compiling to machine code!  Numpy is calling down into machine code too for the inner loops, so there's no inherent reason for one to be faster.  But Numpy's hands are tied when it comes to making intermediate arrays and effectively looping many times, so Numba wins here.